### ISSUE TITLE
Question on Figure 6 Calculation and 1-bit Mask Storage Efficiency

### ISSUE THREAD

**Issue by duludulu-daladala** [harveyhuang18/EMR_Merging#9]
Thank you for your great work! I have a question regarding the memory efficiency analysis in your paper.

In the paper, it is mentioned that the mask size is 1-bit. However, in PyTorch, there is no native 1-bit tensor type, and the bool type actually takes 1 byte (8 bits) per element. And according to your code, you used the bool type of tensor. This means that, in practice, the memory saving is only about 2x to 4x compared to float16 or float32, rather than an 16x reduction.

Given this, I have two questions:

1. How was Figure 6 calculated? Were the memory savings based on an assumed 1-bit storage, or was there some compression technique used?

2. Under these conditions, does the proposed merging method still provide significant advantages in terms of memory efficiency?

I appreciate your time and look forward to your insights!

Best regards,

**Comment by harveyhuang18** [comment]
Sorry for the late response and thanks for your attention to our work.

In our code, the masks are set to bool tensors for convenience, which is not the optimal solution. However, theoretically and technically, the masks can be set to 1-bit. For example, you can try to use ***np.packbits*** to compress the masks before storing them. We tried this on merging ViT-B/32 models and found that the stored mask for a task is 14.2MB while the storage cost for a model is 453.9MB, which means the storage cost can be indeed sharply reduced to 1/32.

We provide the code here for compressing the masks, please try it. In ***merge_vit/src/main_emr_merging.py***, after the merging process:

```
# merge models
vector_unified, masks, rescalers = emr_merge(task_vectors)

from ties_merging_utils import state_dict_to_vector
# store a mask
mask = {}
for n in masks:
    # store the mask for task 0
    mask[n] = masks[n][0] 
mask = state_dict_to_vector(mask)
packed_tensor = np.packbits(mask.numpy())
np.save('./packed_mask.npy', packed_tensor)
```

Hopefully this could solve your problem. If you have any other problems, feel free to continue contacting us.

**Comment by nathanielyvo** [comment]
> Sorry for the late response and thanks for your attention to our work.
> 
> In our code, the masks are set to bool tensors for convenience, which is not the optimal solution. However, theoretically and technically, the masks can be set to 1-bit. For example, you can try to use _**np.packbits**_ to compress the masks before storing them. We tried this on merging ViT-B/32 models and found that the stored mask for a task is 14.2MB while the storage cost for a model is 453.9MB, which means the storage cost can be indeed sharply reduced to 1/32.
> 
> We provide the code here for compressing the masks, please try it. In _**merge_vit/src/main_emr_merging.py**_, after the merging process:
> 
> ```
> # merge models
> vector_unified, masks, rescalers = emr_merge(task_vectors)
> 
> from ties_merging_utils import state_dict_to_vector
> # store a mask
> mask = {}
> for n in masks:
>     # store the mask for task 0
>     mask[n] = masks[n][0] 
> mask = state_dict_to_vector(mask)
> packed_tensor = np.packbits(mask.numpy())
> np.save('./packed_mask.npy', packed_tensor)
> ```
> 
> Hopefully this could solve your problem. If you have any other problems, feel free to continue contacting us.


I have some concerns regarding the practical impact on memory efficiency:

1. While np.packbits reduces storage size, the masks still need to be converted back into torch.Tensor for computation, meaning the GPU memory usage remains unchanged. The primary goal is reducing GPU memory usage, not only storage. Since the packed masks must be unpacked during inference, there may no significant reduction in peak GPU memory consumption.
2.  If the method does not reduce memory usage during inference, whether its real-world applicability is limited? especially for large models where GPU memory is the main bottleneck.

**Comment by harveyhuang18** [comment]
Your concern is reasonable. However, the whole merging process of EMR-Merging could be finished on a CPU. Loading the models to GPU after applying the masks and rescalers does not increase the GPU memory.

p.s. Reading your WUDI-Merging, interesting work!

**Comment by nathanielyvo** [comment]
> Your concern is reasonable. However, the whole merging process of EMR-Merging could be finished on a CPU. Loading the models to GPU after applying the masks and rescalers does not increase the GPU memory.
> 
> p.s. Reading your WUDI-Merging, interesting work!

Thank you for your detailed explanation! That makes a lot of sense. I appreciate your insights on the GPU memory usage. Its a very nice job!

**Comment by harveyhuang18** [comment]
Thanks for your insightful question and appreciation. This means a lot to us.
